DIPs/DIP1009.md

```diff
@@ -23,18 +23,45 @@ D has already made a significant commitment to the theory of [Design by Contract
+}
+```
+The second reason contracts are hard to use is that they require the use of the keyword `do` (or `body`, see [DIP1003](https://github.com/dlang/DIPs/blob/master/DIPs/DIP1003.md)) before the function body. Any way of writing them that could eliminate this need would improve things.
+
+Finally, contracts in D require explicitly writing `assert`, when the vast majority of programmers can be expected to use D's existing `assert` functionality. It is unnecessary to force programmers to write `assert` when it can be assumed. By keeping all existing syntaxes, programmers who need to write something other than `assert` will still have that option available.
+
+All these limitations can be removed by adding a new syntax. History shows that a small change in syntax can dramatically change programmers' willingness to use language features. An increase in usage of D's contracts can be expected.
+
+Furthermore, because the existing `invariant` grammar is so similar to `in` and `out` contract grammar, any improvements to `in` and `out` contracts can be transferred easily to `invariant`s as well. Consistency across language constructs is desirable.
 
 ## Description
 
-Allow `in`, `out`, and `invariant` contracts to be written as parenthesized expressions. Allow multiple expressions of this kind. Create a new grammatical construct called ContractParameters, which copies and replaces the parameters to `assert` expressions. Use the new construct for `in`, `out`, `assert`, and `invariant` expressions alike. In the case of `out` expressions, require a semicolon separating an optional return identifier from the ContractParameters, which come afterward. `invariant` expressions require a semicolon terminator. The keyword `do` or `body` is not required after `in` and `out` contracts that use the new syntax. For example:
+Allow `in`, `out`, and `invariant` contracts to be written as parenthesized expressions. Allow multiple expressions of this kind. Create a new grammatical construct called ContractParameters, which copies and replaces the parameters to `assert` expressions. Use the new construct for `in`, `out`, `assert`, and `invariant` expressions alike. `in` and `out` expressions must come at the end of the function declarator suffix, and before the regular contracts, if any. `invariant` expressions require a semicolon terminator. The keyword `do` or `body` is not required after the new `in` and `out` expressions. The choice of syntax for `out` expressions requires a separate analysis (see the Analysis section below). For the sake of the following example, the `out` expression syntax requires a semicolon separating an optional return identifier from its subsequent ContractParameters:
 
 ```d
 int fun(ref int a, int b)
-  in(a > 0)
-  in(b >= 0, "b cannot be negative!")
-  out(r; r > 0, "return must be positive")
-  out(; a !=0)
+   in(a > 0)
+   in(b >= 0, "b cannot be negative!")
+   out(r; r > 0, "return must be positive")
+   out(; a !=0)
```

Comment:
Shouldn't the 2nd contract expression `a != 0` be evaluated in a separate scope from the first `r > 0`, as `r` shouldn't be defined for the 2nd contract. This would help prevent bugs where there are name conflicts with other symbols in scope.

In fact, the lowering below is undefined when there are 2 `out` contracts with different `out` variable names: `out (?){...}`. Perhaps use local scoped blocks for each `assert`, with their own local named alias for the `out` variable name, which should be a unique compiler generated identifier.